include/mysqlpp/udf_context.hpp

```diff
@@ -0,0 +1,144 @@
+  }
+
+  ext::string_view get_attribute(std::size_t index) const noexcept {
+    return {args_->attributes[index], args_->attribute_lengths[index]};
+  }
+
+  bool is_arg_nullable(std::size_t index) const noexcept {
+    return args_->maybe_null[index] != 0;
+  }
+
+  bool is_result_nullabale() const noexcept { return initid_->maybe_null; }
+
+  bool is_result_const() const noexcept { return initid_->const_item; }
+
+  std::size_t get_result_max_length() const noexcept {
+    return initid_->max_length;
+  }
+
+  std::size_t get_result_decimals() const noexcept { return initid_->decimals; }
+
+  bool is_result_decimals_not_fixed() const noexcept {
+    return initid_->decimals == DECIMAL_NOT_SPECIFIED;
+  }
+
+  void set_arg_type(std::size_t index, item_result_type type) noexcept {
+    args_->arg_type[index] = type;
+  }
+
+  void mark_arg_nullable(std::size_t index, bool nullable) noexcept {
+    args_->maybe_null[index] = (nullable ? 1 : 0);
+  }
+
+  void mark_result_nullable(bool nullable) noexcept {
+    initid_->maybe_null = nullable;
+  }
+
+  void mark_result_const(bool constant) noexcept {
+    initid_->const_item = constant;
+  }
+
+  void set_result_max_length(std::size_t max_length) noexcept {
+    initid_->max_length = max_length;
+  }
+
+  void set_result_decimals(std::size_t decimals) noexcept {
+    initid_->decimals = decimals;
+  }
+
+  void set_result_decimals_not_fixed() noexcept {
+    initid_->decimals = DECIMAL_NOT_SPECIFIED;
+  }
+
+ private:
+  UDF_INIT *initid_;
+  UDF_ARGS *args_;
+
+  template <item_result_type ItemResult>
+  using item_result_tag = std::integral_constant<item_result_type, ItemResult>;
+
+  using string_result_tag = item_result_tag<STRING_RESULT>;
+  using real_result_tag = item_result_tag<REAL_RESULT>;
+  using int_result_tag = item_result_tag<INT_RESULT>;
+  using decimal_result_tag = item_result_tag<DECIMAL_RESULT>;
+
+  udf_context(UDF_INIT *initid, UDF_ARGS *args) noexcept
+      : initid_{initid}, args_{args} {}
+
+  auto get_arg_impl(std::size_t index, string_result_tag) const noexcept {
+    return udf_arg_t<STRING_RESULT>{args_->args[index], args_->lengths[index]};
+  }
+  auto get_arg_impl(std::size_t index, real_result_tag) const noexcept {
+    using result_type = udf_arg_t<REAL_RESULT>;
+    if (args_->args[index] == nullptr) return result_type{};
+    result_type::value_type res;
+    std::memcpy(&res, args_->args[index], sizeof res);
+    return result_type{res};
+  }
+  auto get_arg_impl(std::size_t index, int_result_tag) const noexcept {
+    using result_type = udf_arg_t<INT_RESULT>;
+    if (args_->args[index] == nullptr) return result_type{};
+    result_type::value_type res;
+    std::memcpy(&res, args_->args[index], sizeof res);
+    return result_type{res};
+  }
+  auto get_arg_impl(std::size_t index, decimal_result_tag) const noexcept {
+    return udf_arg_t<DECIMAL_RESULT>{args_->args[index], args_->lengths[index]};
```

Comment:
Do we need `args_->args[index] == nullptr` checkup and returning empty result here? Same question for `string_result_tag` above.

Reply:
For `STRING_RESULT` and `DECIMAL_RESULT` `get_arg_impl()` specializations will return `ext::string_view`.
This is how we can distinguish NULL, empty, non-empty string values
```c++
if (arg.data() == nullptr) {
 // NULL is passed
} else if (arg.empty()) {
  // Empty string "" is passed
} else {
  // Non-empty string is passed
}
```